File: 3909-minimum-increments-to-equalize-leaf-paths/3909-minimum-increments-to-equalize-leaf-paths.java
class Solution {
        int res = 0;
    List<List<Integer>> G;

    public int minIncrease(int n, int[][] edges, int[] cost) {
        G = new ArrayList<>();
        for (int i = 0; i < n; i++) G.add(new ArrayList<>());
        for (int[] e : edges) {
            int u = e[0], v = e[1];
            G.get(u).add(v);
            G.get(v).add(u);
        }
        dfs(0, -1, cost);
        return res;
    }

    private long dfs(int i, int f, int[] cost) {
        List<Long> score = new ArrayList<>();
        long ma = 0;
        for (int j : G.get(i)) {
            if (j == f) continue;
            long val = dfs(j, i, cost);
            if (val > ma) ma = val;
            score.add(val);
        }
        if (score.isEmpty()) {
            return cost[i];
        }
        for (long val : score)
            if (val < ma)
                res++;
        return ma + cost[i];
    }
}

File: 3909-minimum-increments-to-equalize-leaf-paths/3909-minimum-increments-to-equalize-leaf-paths.cpp
class Solution {
    vector<vector<int>> adj;
    int result = 0;
    // DFS
    long long dfs(int cur, int parent, vector<int>& costs){
        // post order
        vector<long long> child;
        for(auto next : adj[cur]){
            if(next != parent){
                child.push_back(dfs(next, cur, costs));
            }
        }
        long long maxChild = 0;
        if(child.size() > 0){
            maxChild = *max_element(child.begin(), child.end());
            for(auto val : child){
                if(val != maxChild){ // need to add subtree root value
                    ++result; 
                }
            }
        }
        return maxChild + costs[cur];
    }
public:
    int minIncrease(int n, vector<vector<int>>& edges, vector<int>& costs) {
        // create graph
        adj.resize(n);
        for(auto& edge : edges){
            adj[edge[0]].push_back(edge[1]);
            adj[edge[1]].push_back(edge[0]);
        }
        // run DFS for each subtree node, it cost need to be the same
        dfs(0, -1, costs);
        return result;
    }
};